core/core/test/serverspec/spec/kafka/kafka_spec.rb

```diff
@@ -1,28 +1,118 @@
 require 'spec_helper'
 
-
-describe service('kafka') do
-  it { should be_enabled }
-  it { should be_running }
+describe 'Checking if kafka service is running' do
+  describe service('kafka') do
+    it { should be_enabled }
+    it { should be_running }
+  end
 end
 
-describe port(9092) do
-  it { should be_listening }
+describe 'Checking if kafka user exists' do
+  describe group('kafka') do
+    it { should exist }
+  end
+  describe user('kafka') do
+    it { should exist }
+    it { should belong_to_group 'kafka' }
+    it { should have_home_directory '/home/kafka' }
+    it { should have_login_shell '/usr/sbin/nologin' }
+  end
 end
 
-describe port(7071) do
-  it { should be_listening }
+describe 'Checking if the ports are open' do
+  describe port(9092) do
+    let(:disable_sudo) { false }
+    it { should be_listening }
+  end
+end  
+
+describe 'Listing down all the active brokers' do
+  describe command('echo "ls /brokers/ids" | /opt/kafka/bin/zookeeper-shell.sh localhost:2181') do
+    its(:stdout) { should match /Welcome to ZooKeeper!/ }
+    its(:stdout) { should match /\[(\d+(\,\s)?)+\]/ }  # pattern: [0, 1, 2, 3 ...]
+    its(:exit_status) { should eq 0 }
+  end
 end
 
-describe user('kafka') do
-  it { should exist }
-  it { should have_login_shell '/usr/sbin/nologin' }
-  it { should belong_to_group 'kafka' }
-  it { should belong_to_group 'jmx-exporter' }
+describe 'Checking if the number of kafka brokers is the same as indicated in the inventory file' do
+  describe command('echo "dump" | curl -v "telnet://localhost:2181" | grep -c brokers | tr -d "\n"') do
+    it "is expected to be equal" do
+    expect(subject.stdout.to_i).to eq count_inventory_roles("kafka")
+    end
+  end
 end
 
-#check if output from jmx exporter is correct
-describe command('curl -s localhost:7071 | grep kafka') do
-  its(:stdout) { should match /kafka_/ }
-  its(:exit_status) { should eq 0 }
+describe 'Checking the possibility of creating a topic, producing and consuming messages' do
+
+  kafka_brokers_count = count_inventory_roles("kafka")
+  timestamp = DateTime.now.to_time.to_i.to_s
+  topic_name = 'topic' + timestamp
+  partitions = kafka_brokers_count*3
+  message = 'test message 1'
+  message2 = 'test message 2'
+  message3 = 'test message 3'
+
+  describe 'Checking if the topic was created' do
+    describe command("/opt/kafka/bin/kafka-topics.sh --create --zookeeper localhost:2181 --replication-factor #{kafka_brokers_count} \
+    --partitions #{partitions} --topic #{topic_name}") do
+      its(:stdout) { should match /Created topic "#{topic_name}"./ }
+      its(:exit_status) { should eq 0 }
+    end
+  end
+
+  describe 'Starting consumer process' do
+    describe command("/opt/kafka/bin/kafka-console-consumer.sh --bootstrap-server localhost:9092 --topic #{topic_name} >> /tmp/#{topic_name}.txt 2>&1 &") do
+      its(:exit_status) { should eq 0 }
+    end
+  end
+
+  describe 'Waiting 5 sec for consumer process to be started' do 
```

Comment:
It looks like this tests `sleep` command. Is it possible to check the status dynamically instead waiting for 5 seconds - maybe 3 seconds would be enough or 5 not always enough. I'm thinking about a loop for checking the status in e.g. 1 second interval with a timeout (e.g. 5 sec or more).